clients/src/test/java/org/apache/kafka/common/network/SslSelectorTest.java

```diff
@@ -224,4 +233,42 @@ private SslSender createSender(InetSocketAddress serverAddress, byte[] payload)
         return new SslSender(serverAddress, payload);
     }
 
+    private static class TestSslChannelBuilder extends SslChannelBuilder {
+
+        public TestSslChannelBuilder(Mode mode) {
+            super(mode);
+        }
+
+        @Override
+        protected SslTransportLayer buildTransportLayer(SslFactory sslFactory, String id, SelectionKey key, String host) throws IOException {
+            SocketChannel socketChannel = (SocketChannel) key.channel();
+            SSLEngine sslEngine = sslFactory.createSslEngine(host, socketChannel.socket().getPort());
+            TestSslTransportLayer transportLayer = new TestSslTransportLayer(id, key, sslEngine);
+            transportLayer.startHandshake();
+            return transportLayer;
+        }
+
+        /*
+         * TestSslTransportLayer will read from socket once every two tries. This increases
+         * the chance that there will be bytes buffered in the transport layer after read().
+         */
+        class TestSslTransportLayer extends SslTransportLayer {
+            boolean muteSocket = false;
```

Comment:
This seems to work. I ran the test case 50 times locally without failure. Wondering though if we could do something simpler. Like maybe we could override `hasBytesBuffered` to always return true? I think that would be sufficient for this test case and we wouldn't need to worry about this becoming another transient failure case.

Reply:
@hachikuji Thanks for the comment. I understand your concern that the existing test may become a transient failure because we can not be confident there will be bytes buffered in the transport layer.

On the other hand, if we simply override `hasBytesBuffered` to always return true, then it seems equivalent to not waiting for bytes to be buffered in the `testDisconnectWithIntermediateBufferedBytes()`. In this case, the test will continue to validate that the selector is empty even though it is actually already empty before the selector is closed. Thus overriding the `readFromSocketChannel()` seems like a better solution to make this test useful. Not sure I understand this test correctly. What do you think?

Reply:
Pretty sure it would be sufficient to test removal from `Selector.keysWithBufferedRead`, which was its initial purpose. We can leave it as is if you prefer. I was just concerned that the test case might be flaky. In fairness, I can't say this is worse than the current situation. We can consider a proper mock if the test case proves unreliable.

Reply:
@hachikuji Given that you are sure about it and that you have better understanding on this test, I was actually going to override the `keysWithBufferedRead` as suggested. Anyway, since this patch has been merged, we can override the `keysWithBufferedRead` as you suggested if there is any transient failure with this test in the future.

Thanks for all the comments.

Reply:
Yeah, figured I'd just merge it since it's annoying blocking on an issue like this. If the test case proves unreliable, we have some options.